(#773) Add new Chocolatey CLI dependency troubles page (#774)
chocolatey/docs: input/en-us/choco/troubleshooting/dependency-troubles.md

---

## `choco upgrade all` Fails to Upgrade a Package, Which Cascades into Failing Many More Seemingly Unrelated Packages

Due to the way dependency resolution works, packages with certain dependency specifications can occasionally trigger undesirable behaviour when using `choco upgrade all` if the dependencies fail to install.
Typically this will be encountered with meta-packages, such as `obs-studio` which takes a dependency on `obs-studio.install`.
When running `choco upgrade all` it will attempt to upgrade every installed package. Any packages which fail to install may be retried as part of a later dependency chain.
For example, `obs-studio.install` may be upgraded, fail, and then later be tried again as part of the dependency chain for `obs-studio` itself.
As a result, Chocolatey CLI's packages may be in an inconsistent state.
This is typically characterised by a dependency resolution error message when performing package operations:

```cmd
Unable to resolve dependencies. 'obs-studio.install 29.0.2' is not compatible with 'obs-studio 29.1.1 constraint: obs-studio.install (>= 29.1.1)'.
```

Typically this is resolvable by directly upgrading the affected dependency.

> This is normally resolved by upgrading the affected dependency.

Reviewer: Repeated sentence here

To resolve the issue, we run `choco upgrade obs-studio.install` to get the correct dependency version installed.
For packages that have a specific version dependency, add the `--version` option and specify the exact version to install.

![Partial output from choco upgrade all, showing that obs-studio.install is attempted and failed](/assets/images/chocolatey/dependency-resolution-failure-1.png)
![Continued output from choco upgrade all, showing that obs-studio metapackage is installed and subsequently re-attempts the upgrade to obs-studio.install which again fails, followed by a dependency resolution error in the next package and failing a seemingly unrelated package installation](/assets/images/chocolatey/dependency-resolution-failure-2.png)

### Example 2

> :choco-warning: **WARNING**
>
> This example deals with multiple missing dependencies, and installs them while ignoring dependencies.
> It is recommended to ensure you're installing the proper versions needed for your dependencies.

> The follow scenario illustrates what can happen during an `install` where multiple existing packages have dependencies that are not currently met.

Reviewer: ```suggestion
The following scenario illustrates what can happen during an `install` where multiple existing packages have dependencies that are not currently met.

```

Due to the dependency version range specified by `obs-studio`, the version of `obs-studio.install` that remains installed (due to the upgrade failure) does not satisfy the dependency requirement.

To resolve the issue, we run `choco upgrade obs-studio.install` to get the correct dependency version installed.
For packages that have a specific version dependency, add the `--version` option and specify the exact version to install.

![Partial output from choco upgrade all, showing that obs-studio.install is attempted and failed](/assets/images/chocolatey/dependency-resolution-failure-1.png)
![Continued output from choco upgrade all, showing that obs-studio metapackage is installed and subsequently re-attempts the upgrade to obs-studio.install which again fails, followed by a dependency resolution error in the next package and failing a seemingly unrelated package installation](/assets/images/chocolatey/dependency-resolution-failure-2.png)

### Example 2 - Installing a Package Fails Due to Existing Missing Dependencies

> :choco-warning: **WARNING**
>
> This example deals with multiple missing dependencies, and installs them while ignoring dependencies.

> > It is recommended to ensure you're installing the proper versions needed for your dependencies.

Reviewer: ```suggestion
> You will need to ensure you're installing the correct versions required for your dependencies.
```
I think this could stand to be worded more firmly, since if they don't they're just gonna get more problems lol

  ![Output from choco install KB2919355 where it fails to resolve dependency 'KB2999226'](/assets/images/chocolatey/kb-dependency-failure-3.png)

4. Now with `KB2919355` installed we continue with `choco install KB2999226`.
Again we make note of the next package (`KB2919442`) and install this package while ignoring dependencies: `choco install KB2999226 --ignore-dependencies`.

  ![Output from choco install KB2999226 where it fails to resolve dependency 'KB2919442'](/assets/images/chocolatey/kb-dependency-failure-4.png)

5. Now with `KB2999226` installed, we attempt to run `choco install KB2919442`.
We see that this one is actually already installed, so we attempt to install `glab` again.
When `glab` fails to install, we make note of the next package missing in our chain: `chocolatey-windowsupdate.extension`.

  ![Output from choco install KB2919442 where it is already installed. Following output is from choco install glab where it fails to resolve dependency 'chocolatey-windowsupdate.extension'](/assets/images/chocolatey/kb-dependency-failure-5.png)

> 6. Installing this in the same ways as before (`choco install chocolatey-windowsupdate.extension`, then `choco install chocolatey-windowsupdate.extension --ignore-dependencies`).

Reviewer: ```suggestion
6. We install this in the same way as before (`choco install chocolatey-windowsupdate.extension`, then `choco install chocolatey-windowsupdate.extension --ignore-dependencies`).
```

Keeping consistent... tense? framing? as with previous steps
